core/src/main/scala/org/apache/spark/util/collection/ExternalSorter.scala

```diff
@@ -306,14 +297,10 @@ private[spark] class ExternalSorter[K, V, C](
 
     var success = false
     try {
-      val it = collection.destructiveSortedIterator(partitionKeyComparator)
+      val it = collection.destructiveSortedWritablePartitionedIterator(comparator)
       while (it.hasNext) {
-        val elem = it.next()
-        val partitionId = elem._1._1
-        val key = elem._1._2
-        val value = elem._2
-        writer.write(key)
-        writer.write(value)
+        val partitionId = it.nextPartition()
+        it.writeNext(writer)
         elementsPerPartition(partitionId) += 1
         objectsWritten += 1
```

Comment:
Is this still correct?  Based on my understanding of `destructiveSortedWritablePartitionedIterator` (which is kind of a mouthful!), it looks like the `it.writeNext(writer)` call should be writing an entire partition's worth of records to the output, but here and on the previous line we're performing some increments that I think only make sense for a record-at-a-time iterator, which we're no longer using in this block.